### PR TITLE
feat(wrangler): add container support to worker previews

### PR DESCRIPTION
**Pull request by WillTaylorDev**
Worker previews can't run containers today. This adds a `previews.containers` configuration block
so a preview can declare the containers it needs, and wrangler provisions them when it deploys.

This replaces #13781, which did the same provisioning work but also deleted the container
applications from the client on `wrangler preview delete`. Greg pointed out in review that a preview
should tear down the resources it owns at the API level, and that the Builds API auto-deletes
previews without ever invoking wrangler, so anything relying on the client would orphan running
containers. The config service is taking on teardown instead, which also removes a failure mode the
client side version had: when Cloudchamber was unreachable, wrangler had to refuse to delete the
preview at all rather than risk losing the slug that named the applications.

The provisioning side is unchanged in intent from #13781 and carries the review fixes from it.
Container names are generated as `{worker_name}_{preview_slug}_{class_name}`, matching the server
side `PreviewNamer` so the control plane can find them. Bindings that name a Durable Object
implemented by another Worker through `script_name` are skipped, since that Worker owns its own
container application. Configuration errors are caught before the preview deployment is created, so
a typo fails loudly instead of leaving a preview running with no container.

Max wrote the original feature commit and it's preserved here with his authorship.

```jsonc
// wrangler.json
{
  "previews": {
    "durable_objects": {
      "bindings": [{ "name": "MY_CONTAINER", "class_name": "MyContainer" }]
    },
    "containers": [
      {
        "class_name": "MyContainer",
        "image": "registry.cloudflare.com/example:latest"
      }
    ]
  }
}
```

---

- Tests
  - [x] Tests included/updated
  - [ ] Automated tests not possible - manual testing has been completed as follows:
  - [ ] Additional testing not necessary because:
- Public documentation
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: private beta

### CODE REVIEW

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- @cloudflare/cloudchamber
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>

- packages/wrangler/src/containers/deploy.ts: [@cloudflare/cloudchamber]

</details>

File: packages/workers-utils/tests/config/validation/normalize-and-validate-config.test.ts [hunk 1]
```diff
@@ -11427,6 +11427,169 @@ describe("normalizeAndValidateConfig()", () => {
...
+							},
+						],
+					},
+				} as unknown as RawConfig;
+
+				const { diagnostics } = normalizeAndValidateConfig(
+					rawConfig,
+					undefined,
+					undefined,
+					{ env: undefined }
```
Comment (GregBrimble) [inline packages/workers-utils/tests/config/validation/normalize-and-validate-config.test.ts:12581]: Add a test which asserts that worker name being dynamically passed in (e.g. through `npx wrangler deploy --name` or `npx wrangler preview --worker-name`) still works.

File: packages/workers-utils/src/config/validation.ts [hunk 2]
```diff
@@ -3375,11 +3375,75 @@ const validateBindingArray =
...
+	return (diagnostics, field, value, config) => {
+		if (
+			Array.isArray(value) &&
+			value.some(
+				(entry) =>
+					entry && typeof entry === "object" && entry.name !== undefined
+			)
+		) {
+			diagnostics.errors.push(
+				`"name" is not allowed on "${field}" entries; preview container application names are auto-generated by wrangler from the worker name, preview slug, and class name.`
```
Comment (GregBrimble) [inline packages/workers-utils/src/config/validation.ts]: This name error stuff surprises me a bit, but I'll keep reading and see where I land on it. Just noting it to myself.

File: packages/deploy-helpers/src/preview/preview.ts [hunk 3]
```diff
@@ -78,6 +88,181 @@ export type PreviewResult = {
...
+		) {
+			continue;
+		}
+		throw new UserError(
+			`The container class_name "${container.class_name}" in "previews.containers" does not match any Durable Object binding in "previews.durable_objects". Container config is not inherited from the top-level config, so the Durable Object must be declared under "previews.durable_objects" in your ${configFileName(config.configPath)} file.`,
+			{
+				telemetryMessage: "no preview DO binding matches container class_name",
+			}
+		);
+	}
```
Comment (GregBrimble) [inline packages/deploy-helpers/src/preview/preview.ts:188]: Feels brittle and possibly even actively outdated with `ctx.exports`? Is this what normal Containers do today? They require an explicit binding? If yes, fine with it. But they probably want to untangle this soon-ish too.

File: packages/deploy-helpers/src/preview/preview.ts [hunk 4]
```diff
@@ -78,6 +88,181 @@ export type PreviewResult = {
...
+	// match carries `script_name`. A class bound both locally and cross-script
+	// would then fail as though another Worker owned it, purely because of
+	// binding order. Put the locally implemented bindings first so the lookup
+	// lands on the one this preview owns.
+	const localBindingsFirst = [
+		...previewDOBindings.filter((b) => b.script_name === undefined),
+		...previewDOBindings.filter((b) => b.script_name !== undefined),
+	];
+
+	const observability = previews?.observability ?? config.observability;
```
Comment (GregBrimble) [inline packages/deploy-helpers/src/preview/preview.ts:212]: Just obs? Why is obs special here? What about other inheritable keys e.g. logpush, limits and cache?

**Review by GregBrimble — APPROVED** [review]
Okay, yeah, saw nothing in here that specifically meant (to my eyes) that we can't and shouldn't support user-provided names, so my question around that stands. But it's purely additive, so won't block right now.

File: packages/workers-utils/src/config/validation.ts (hunk 2 above)
Comment (GregBrimble) [reply packages/workers-utils/src/config/validation.ts]: Yeah, it seems to me that the creation flow would support customizing the name (we just use that new name if provided). And now that teardown is left to the server, (and it trickles strongly through the DO namespace), then there's no reason to lookup-by-name or anything, so I can't imagine why we'd need this deterministic name for the deletion flow either. Consider removing this restriction?

Comment (WillTaylorDev) [reply packages/workers-utils/src/config/validation.ts]: Ahhh, yeah looking at this closer you're right. I thought the generated name had to match exactly. I'll chop this!

File: packages/deploy-helpers/src/preview/preview.ts (hunk 4 above)
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/preview.ts:212]: It's not inheritance (the docblock does a bad job of conveying this lol, I'll fix)

The container application has its own observability field, and we read the current obs settings onto it ( configuration.observability.logs.enabled)

File: packages/deploy-helpers/src/preview/preview.ts (hunk 3 above)
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/preview.ts:188]: Oop, they don't. Deploy resolves the class from migrations and exports as well as bindings, so yeah looks like ctx.exports only would work for that.

I'll swap in the same helper deploy uses. And agreed on untangling, the DO folks are landing on container config (and what in the wrangler config owns that), so I'll avoid any opinionated approaches otherwise 😅

File: packages/deploy-helpers/src/preview/preview.ts (hunk 4 above)
Comment (GregBrimble) [reply packages/deploy-helpers/src/preview/preview.ts:212]: Ohh! I missed that entirely. Okay, absolutely fair enough then. If we're matching the functionality of production container applications, and imposing our existing pattern for obs inheritance on that, then I think that's good with me.

Might be good to have a test for this specifically too, just to lock that in and illustrate it.

File: packages/deploy-helpers/src/preview/preview.ts (hunk 3 above)
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/preview.ts:188]: So this now mirrors what wrangler deploy already does for unbound DOs (falls back to the namespaces list). One difference though, deploy matches on class + script, but for previews, script is the parent worker's name across all of its previews, so I match on the preview id to avoid grabbing the parent's namespace or a sibling preview's. (there's a test for that)

File: packages/deploy-helpers/src/preview/preview.ts [hunk 5]
```diff
@@ -78,6 +88,183 @@ export type PreviewResult = {
...
+	}
+
+	const filteredContainers = previewContainers
+		.filter((c) => ownedDOClasses.has(c.class_name))
+		.map((c) => ({
+			...c,
+			name:
+				c.name ??
+				previewContainerAppName(parentWorkerName, previewSlug, c.class_name),
+		}));
```
[the remark this answers is not in the document]
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/preview.ts]: The name isn't what associates a preview's containers with the preview, so a custom name doesn't orphan anything.

PreviewNamer in the config service names Durable Object namespaces, not Cloudchamber applications. A container application is created bound to a durable_object_namespace_id, and a preview's namespaces are preview scoped, so the backend can find them without parsing the application name. Teardown is handled backend side like we already said.

The test comment you quoted made was making an inccorect claim, though, so I've removed it.

File: packages/wrangler/src/containers/deploy.ts [hunk 6]
```diff
@@ -190,14 +190,20 @@ async function fetchUploadedVersion(
...
 	class: string;
 	name: string;
 	script: string;
 	useSqlite: boolean;
+	/**
+	 * Set when the namespace belongs to a Worker preview. For those, `script` is
+	 * the parent Worker's name, so `preview.id` is what distinguishes a
+	 * preview's namespace from the parent's and from other previews'.
+	 */
+	preview?: { id: string; slug: string; name: string };
```
[the remark this answers is not in the document]
Comment (WillTaylorDev) [reply packages/wrangler/src/containers/deploy.ts:205]: Smells like a real bug to me, but predates this PR and sits on the deploy path (not a path I'm modifying), so I'd rather not widen the scope here. I'd prefer both the preview and deploy path to be equivalent as much as possible. We have some followup tickets cut for preventing any sort of cross preview usage in non preview workers, which should handle this.

**Review by jamesopstad — COMMENTED** [review]
I've reposted some comments from the Devin review. If they don't require a change then just a comment is fine.

File: packages/workers-utils/src/config/validation.ts
[the remark this answers is not in the document]
Comment (WillTaylorDev) [reply packages/workers-utils/src/config/validation.ts:1]: Does look like dead code, but predates this MR and touches the deploy path. I'm trying to keep a small blast radius here. Would be good as a followup, but ideally would kick out of this MR.

File: packages/deploy-helpers/src/preview/preview.ts [hunk 7]
```diff
@@ -78,6 +88,183 @@ export type PreviewResult = {
...
+	// `getNormalizedContainerOptions` resolves a container's Durable Object with
+	// `find()` on `class_name`, and rejects the container outright if that first
+	// match carries `script_name`. A class bound both locally and cross-script
+	// would then fail as though another Worker owned it, purely because of
+	// binding order. Put the locally implemented bindings first so the lookup
+	// lands on the one this preview owns.
+	const localBindingsFirst = [
+		...previewDOBindings.filter((b) => b.script_name === undefined),
+		...previewDOBindings.filter((b) => b.script_name !== undefined),
+	];
```
[the remark this answers is not in the document]
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/preview.ts:210]: Agreed on the diagnosis, and it's the same order sensitivity as the other deploy-path finding here. Making the predicate prefer a binding without script_name is the better end state and would make this reorder redundant, but it changes wrangler deploy behaviour, so I'm keeping it out of a preview change. Tracking it with the other deploy-path fix.

Same thing here as my other comment. You're correct, but expands the scope of changes and touches the shared deploy path. Prefer to split out with the other dead code fixups, in a later MR.

File: packages/wrangler/src/preview/containers.ts [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+				deployment.id,
+				false,
+				dockerPath,
+				// `preview()` already verified Docker before creating the
+				// deployment, so skip the redundant per-container check.
+				false
+			);
+		} else {
+			imageRef = { newTag: (container as ImageURIConfig).image_uri };
+		}
```
[the remark this answers is not in the document]
Comment (WillTaylorDev) [reply packages/wrangler/src/preview/containers.ts:162]: Redirecting that subprocess output means touching shared build code on the deploy path, so I'm leaving it. I'll drop the "stays parseable" wording from the changeset though

File: packages/wrangler/src/preview/containers.ts [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+	setLogLevel("error");
+	try {
+		return await runWithLogLevel("error", () =>
+			applyPreviewContainers(
+				scopedConfig,
+				normalisedContainerConfig,
+				deployment,
+				accountId
+			)
+		);
```
[the remark this answers is not in the document]
Comment (WillTaylorDev) [reply packages/wrangler/src/preview/containers.ts:69]: I actually gave this a shot in Max's original MR, and it got super leaky. Suppressing properly means threading a quiet flag through the shared cloudchamber helpers. Acceptable to treat it as a known edge case.

**Review by jamesopstad — APPROVED** [review]
Thanks for resolving the comments. The lock file was temporarily broken by another PR so you'll need another rebase.

File: .changeset/cyan-planets-ring.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
...tainers explicitly in the `previews` block to enable them for previews. This mirrors how `previews.durable_objects` works today. Wrangler names each preview container application `{worker_name}_{preview_slug}_{class_name}`, unless the entry sets its own `name`, which is used as given. A Durable Object class is backed by at most one container application, so the validator rejects two entries that share a `class_name`, or that share an explicit `name`. Wrangler skips container applications bound to Durable Object classes that another Worker implements through `script_name`, because the implementing Worker owns its own container application. A binding is not required: a Durable Object declared through `migrations` or `exports` and reached only over `ctx.exports` can still back a container. Every entry must set `class_name`, even where a Durable Object declared in `exports` names its container instead, because that reference resolves against the top level `containers` array. A `previews.containers` entry whose `class_name` matches no Durable Object class at all is rejected before the preview deployment is created, so a typo fails loudly instead of producing a preview with no container.
```
Comment (petebacondarwin) [inline .changeset/cyan-planets-ring.md]: One concern here (since I rebased on top of the ctx.exports containers support change):

> Every entry must set `class_name`, even where a Durable Object declared in `exports` names its container instead, because that reference resolves against the top level `containers` array. 

This is going in the wrong direction, since the idea of the new exports container support was to flip this round.

I think what this does is highlight a problem with how `exports` are inheritable while `containers` are non-inheritable.

Comment (WillTaylorDev) [reply .changeset/cyan-planets-ring.md]: Yup, good convos in backchannels. Waiting on PM feedback

**Review by petebacondarwin — APPROVED** [review]
With the expectation that we get this working more cleanly in the future, I think this is good to go.

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- ✅ @cloudflare/cloudchamber
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>


</details>

File: packages/deploy-helpers/src/preview/shared.ts [hunk 11]
```diff
@@ -344,6 +345,84 @@ export function extractConfigBindings(config: Config): EnvBindings {
...
+ * Compose the auto-generated container application name for a preview-scoped
+ * container, in the form `{parentWorkerName}_{previewSlug}_{className}`.
+ */
+export function previewContainerAppName(
+	parentWorkerName: string,
+	previewSlug: string,
+	className: string
+): string {
+	return `${parentWorkerName}_${previewSlug}_${className}`;
+}
```
Comment (mattjohnsonpint) [inline packages/deploy-helpers/src/preview/shared.ts:457]: This app name is composed here, but as far as I can tell - it's never validated.  There are strings that could be valid worker names but would results in invalid container app names.

Specifically, the full app name must:

- be <= 253 characters total
- not start or end in a dash (e.g., `"-foo"` and `"bar-"` are invalid)
- not contain two consecutive dashes (e.g., `"foo--bar"` is invalid)
- not start with a leading numeric digit (e.g., `"1myapp"` is invalid)

So for example, `"9-my-worker"` and `"my--worker"` might both be valid worker names but would produce an invalid app name when used in the first part here.

(`""` and `"-"` are also invalid app names, but they couldn't be produced through this path.)


An invalid name might be accepted here, but it would be rejected when sent to the Cloudchamber API.  Best to catch it up front.
You could ask the user to change the name parts to conform, or you could normalize the constructed app name such that it always confirms with the requirement.

File: packages/wrangler/src/preview/containers.ts [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+				dockerPath,
+				// `preview()` already verified Docker before creating the
+				// deployment, so skip the redundant per-container check.
+				false,
+				// Selects the managed registry for the account's compliance
+				// region. Without it the push defaults to the public registry.
+				scopedConfig
+			);
+		} else {
+			imageRef = { newTag: (container as ImageURIConfig).image_uri };
```
Comment (mattjohnsonpint) [inline packages/wrangler/src/preview/containers.ts]: nit:
```suggestion
			imageRef = { newTag: container.image_uri };
```

File: packages/deploy-helpers/src/preview/shared.ts (hunk 11 above)
Comment (mattjohnsonpint) [reply packages/deploy-helpers/src/preview/shared.ts:457]: Separately - is this composed string ever meant to be parsable back into the three parts, or is it just for display? I don't see anywhere today that's parsing it, but if we ever did then we'd also have to keep underscores out of any of the name parts.  If it's just for display or uniqueness, then it's fine. It's just not reversible.  (e.g., `my_worker_slug_45_some_class_name`)

File: packages/deploy-helpers/src/preview/preview.ts [hunk 13]
```diff
@@ -422,6 +668,14 @@ export async function preview(
 		}
 	}
 
+	const { scopedContainerConfig, normalisedContainerConfig } =
+		await prepareContainersForPreview(
+			config,
+			workerName,
+			previewResource.slug,
+			callbacks
+		);
```
Comment (mattjohnsonpint) [inline packages/deploy-helpers/src/preview/preview.ts:675]: This runs after `createPreview` on line 649.  So let's say `prepareContainersForPreview` when it calls `buildPreviewContainerConfig` throws at line 157 with a `UserError` for any reason - I think it would leave an empty preview behind on the account, right?

**Review by mattjohnsonpint — CHANGES_REQUESTED** [review]
Looking pretty great so far!  A few loose ends to tie up.

File: .changeset/cyan-planets-ring.md (hunk 10 above)
Comment (mattjohnsonpint) [reply .changeset/cyan-planets-ring.md]: Just a data point for the PM conversation:

There's a way this actually breaks. `name` is documented as doing double duty: the application name _and_ the handle `exports[Class].container` points at. In `previews.containers` the second half can never fire, since we always require `class_name` and the exports side resolves against top-level `containers` (see `containers.ts` line 49). So `name` only acts as the app name there — and app names are unique per account, so two branches setting the same one collide, with the second failing after its preview is already live. Someone following the field's own docs walks straight into it.

Simplest fix might be to just not accept `name` in `previews.containers` for now — reject it in validation so it fails early. The class name is already unique within a preview, so name isn't buying anything there, and dropping it keeps the door open for whatever you land on for exports/inheritance. Easy to add back once the semantics are settled.

Note, I distilled this from a much more verbose AI finding, and I verified it myself, but it is indeed a bit confusing still. 😅

File: packages/deploy-helpers/src/preview/preview.ts (hunk 13 above)
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/preview.ts:675]: Technically yes, but they also get auto-cleaned up. The problem is we need the preview slug to generate the app name, so I don't think we can avoid this, unfortunately.

File: packages/deploy-helpers/src/preview/shared.ts (hunk 11 above)
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/shared.ts:457]: Nope, just for display, nothing parsing it today like you said.

Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/shared.ts:457]: I think easiest for us to outright deny invalid names. Will see how we handle it for root level containers.

File: .changeset/cyan-planets-ring.md (hunk 10 above)
Comment (WillTaylorDev) [reply .changeset/cyan-planets-ring.md]: +1 I think outright rejecting name is clearer, I'd be in favor. Let me run by @yomna-shousha for the OK

File: packages/deploy-helpers/src/preview/shared.ts (hunk 11 above)
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/shared.ts:457]: Ended up normalizing the name instead. Here's why.. If the user does hit an error with an invalid name the only remedies are renaming your Worker, renaming your Durable Object class (which needs a migration) or rename your git branch. That's way more obnoxious than just normalizing.

Comment (mattjohnsonpint) [reply packages/deploy-helpers/src/preview/shared.ts:457]: Normalizing is a good idea. A couple of things missed though:

- Total length needs to be <= 253 chars.  A simple truncation isn't ideal, because it could strip off the other unique parts. Or perhaps this isn't an issue if the individual parts are already size constrained? I'm not sure if they are or not.
- An edge case: Some of the normalization could create collisions. ie., `my-worker` and `my--worker`.

There are two ways I can think to resolve the second part:
- You could test for collisions against live data, but that adds a round trip so proabably not worth it.
- You could append some kind of hash based on the pre-normalized string. It doesn't have to be strong hashing, but rather should be short and human friendly. A good option might be base36 at 32 bits, such as `_a3f2b8c`. Of course if something like this is already in the preview slug, then maybe I'm overthinking it.

**Review by mattjohnsonpint — APPROVED** [review]

**Review by workers-devprod — APPROVED** [review]
Codeowners reviews satisfied

File: packages/deploy-helpers/src/preview/shared.ts (hunk 11 above)
Comment (WillTaylorDev) [reply packages/deploy-helpers/src/preview/shared.ts:457]: Done! Refactored a bit of the previous alias generation code to pull out what I could.
